plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/OpenShiftServerEditorSection.java

```diff
@@ -35,196 +57,445 @@
+								return ValidationStatus.ok();
+							}
+							return ValidationStatus.cancel("Please choose a project to deploy.");
+						}
+					})
+					.to(BeanProperties.value(OpenShiftServerEditorModel.PROPERTY_DEPLOYPROJECT)
+					.observe(model))
+					.in(dbc);
+		ControlDecorationSupport.create(
+				selectedProjectBinding, SWT.LEFT | SWT.TOP, null, new RequiredControlDecorationUpdater(true));
+
+		// browse projects
+		Button browseProjectsButton = new Button(parent, SWT.NONE);
+		browseProjectsButton.setText("Browse...");
 		GridDataFactory.fillDefaults()
-				.align(SWT.FILL, SWT.FILL).grab(true, false).span(2, 1).applyTo(overrideProjectSettings);
+				.align(SWT.LEFT, SWT.CENTER).hint(100, SWT.DEFAULT)
+				.applyTo(browseProjectsButton);
+		browseProjectsButton.addSelectionListener(onBrowseProjects(model, browseProjectsButton.getShell()));
+	}
 
-		// connection
-		Label connectionLabel = toolkit
-				.createLabel(projectSettingGroup, "Connection:", SWT.NONE);
-		GridDataFactory.fillDefaults().align(SWT.LEFT, SWT.CENTER).applyTo(connectionLabel);
-		connectionText = toolkit.createText(projectSettingGroup, "", SWT.SINGLE | SWT.BORDER);
-		GridDataFactory.fillDefaults().align(SWT.FILL, SWT.FILL).grab(true, false).applyTo(connectionText);
+	private SelectionListener onBrowseProjects(OpenShiftServerEditorModel model, final Shell shell) {
+		return new SelectionAdapter() {
+			@Override
+			public void widgetSelected(SelectionEvent e) {
+				SelectExistingProjectDialog dialog = 
+						new SelectExistingProjectDialog("Select a project to deploy", shell);
+				if (dialog.open() == Dialog.OK) {
+					Object selectedProject = dialog.getFirstResult();
+					if (selectedProject instanceof IProject) {
+						model.setDeployProject((org.eclipse.core.resources.IProject) selectedProject);
+					}
+				}
+			}
+		};
+	}
+
+	private void createSourcePathControls(Composite container, DataBindingContext dbc) {
+		Label sourcePathLabel = new Label(container, SWT.NONE);
+		sourcePathLabel.setText("Source Path: ");
+		GridDataFactory.fillDefaults()
+				.align(SWT.FILL, SWT.CENTER)
+				.applyTo(sourcePathLabel);
 
-		// project name
-		Label projectNameLabel = toolkit.createLabel(
-				projectSettingGroup, "Project:", SWT.NONE);
+		Text sourcePathText = new Text(container, SWT.BORDER | SWT.READ_ONLY);
 		GridDataFactory.fillDefaults()
-				.align(SWT.LEFT, SWT.CENTER).applyTo(projectNameLabel);
-		projectNameText = toolkit.createText(projectSettingGroup, "", SWT.SINGLE | SWT.BORDER);
+				.align(SWT.FILL, SWT.CENTER).grab(true, false)
+				.applyTo(sourcePathText);
+		ValueBindingBuilder
+				.bind(WidgetProperties.text(SWT.Modify).observe(sourcePathText))
+				.validatingAfterConvert(new IValidator() {
+
+					@Override
+					public IStatus validate(Object value) {
+						if (StringUtils.isEmpty(value)) {
+							return ValidationStatus.cancel("Please provide a source path to deploy to the pod.");
+						}
+						String path = (String) value;
```

Comment:
If value is not instance of String (e.g. null), StringUtils.isEmpty(Object) will return false which is not what is expected.
Even if we are 100% sure that non-null value cannot be other than String, this cast is unsafe as a pattern.